### DISCUSSION TITLE
Why does setting "foo" not update "foo_id" before flush?

### DISCUSSION THREAD

**Discussion by matsimko** [sqlalchemy/sqlalchemy#11886] [Usage Questions]
This is the reverse of the case mentioned in docs: [I set the "foo_id" attribute on my instance to "7", but the "foo" attribute is still None - shouldn't it have loaded Foo with id #7?](https://docs.sqlalchemy.org/en/20/faq/sessions.html#i-set-the-foo-id-attribute-on-my-instance-to-7-but-the-foo-attribute-is-still-none-shouldn-t-it-have-loaded-foo-with-id-7). 

For example, the following will fail: 

```py
assert o.foo_id != 7

foo = session.get(Foo, 7)
o.foo = foo

assert o.foo_id == 7
```

One can obviously use `o.foo.id` instead, but not syncing `o.foo_id` seems like an unnecessary source of bugs.

**Comment by zzzeek** [comment]
the process of synchronizing FK ids necessarily needs primary key identifiers to exist which in the most common case of INSERT are not generated until the SQL statements are emitted to the database and the server-generated values are retrieved.   so any cross-object syncing is always done within the flush process.

**Reply by CaselIT** [reply]
Maybe that faq item may clarify also that point?

**Reply by matsimko** [reply]
Isn't my case only about keeping consistency *within* an object by having a one-way binding from a relationship attribute (`foo`) to the corresponding foreign key attribute (`foo_id`) in the same object, whether or not primary keys exist for any object in question?

**Reply by zzzeek** [reply]
The "foo_id" attribute has to be set to a value, that comes from the other object.  if "foo" was not yet INSERTed, it will have no primary key.   

What you're getting at is "if foo *was* INSERTed already, then we know the value right?"   The answer is "usually".    what if we do this?

```py

foo = session.get(Foo, 7)

foo.id = 12

o.foo = foo

foo.id = 19

assert o.foo_id == ???    # does this keep changing every time?  
# even if there are a few thousand "o"s in memory with this link?


```

By suggesting "foo_id" should be linked to an event on "foo", that means every time an attribute is touched on any object that is referred to by foreign keys elsewhere, we need to constantly update it.  This introduces complexity, bugs, and performance overhead.   it would also be super confusing because this whole "set foo_id on event" only happens if "foo" is persistent (to be updated), not pending (to be inserted).  Having multiple codepaths to accomplish the same goal in different ways at different times leads to a much less predictable library for people trying to use it and makes testing and maintenance of the library itself a lot more difficult.   So we try to keep things simple.

**Reply by matsimko** [reply]
I had in mind something closer to this:

```py
@property
def foo_id(self):
    if self.foo_set_after_foo_id:
        return self.foo.id if self.foo is not None else None
    else:
        return self._foo_id

@foo_id.setter
def foo_id(self, value):
    self.foo_set_after_foo_id = False
    self._foo_id = value

@foo.setter
def foo(self, value):
    self.foo_set_after_foo_id = True
    self._foo = value
```

```py
assert o.foo_id is None and o.foo is None

foo = session.get(Foo, 7)
o.foo = foo
assert o.foo_id == 7

o.foo_id = 4
assert o.foo_id == 4

o.foo = None
assert o.foo_id is None

o.foo_id = 5
assert o.foo_id == 5

assert new_foo.id is None
o.foo = new_foo
assert o.foo_id is None
```

Here, "foo" is allowed to be in a pending state, and no "o" has to be updated after a change to foo's id.

**Reply by zzzeek** [reply]
sure, that's how you'd do that, more likely you'd want to use [AttributeEvents](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.AttributeEvents) which takes advantage of the existing attribute mechanics.

feel free to go with this, it's just not something SQLAlchemy can do automatically

**Reply by matsimko** [reply]
In the end it is probably best to use just `o.foo` and avoid touching `o.foo_id` unless there is a reason to. (If you ever assign different values to both attributes in a single transaction, the outcome after saving to db is somewhat unpredictable).

**Reply by CaselIT** [reply]
> If you ever assign different values to both attributes in a single transaction, the outcome after saving to db is somewhat unpredictable

the relationship value should take precedence in all cases

**Reply by matsimko** [reply]
> the relationship value should take precedence in all cases

It doesn't when the relationship value matches the current value in db:

```py
# in db: o.foo_id == 7
o = session.get(Obj, 1)
foo = session.get(Foo, 7)
o.foo_id = 8
o.foo = foo
session.flush(foo)
o.foo.id == 7  # true
session.refresh(foo)
o.foo.id == 8  # true
```

I guess the assignment to `o.foo` is not detected as a change, and so `o.foo_id` is the only attribute with an actual change and gets written to db.

**Reply by CaselIT** [reply]
that seems likely. IIRC the set event is ignored if the value does not change.
This avoids sql updates that would be pointless, but I guess cause this side effect
